File: 81-search-in-rotated-sorted-array-II.cpp
bool search(vector<int>& nums, int target) {
    int l = 0, r = nums.size() - 1;
    int mid;
    while (l <= r) {
        mid = l + (r - l) / 2;
        if (nums[mid] == target) return true;
        if (nums[mid] == nums[l]) ++l;
        else if (nums[mid] == nums[r]) --r;
        else if (nums[mid] < nums[r]) {
            if (target > nums[mid] && target <= nums[r]) 
                l = mid + 1;
            else 
                r = mid - 1;
        }
        else {
            if (nums[mid] > target && target >= nums[l])
                r = mid - 1;
            else 
                l = mid + 1;
        }
    }
    return false;
}

File: python version/81-search-in-rotated-sorted-array-II.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        left, right = 0, len(nums)-1
        while left < right:
            mid = left + (right-left)//2
            if nums[mid] == target:
                return True
            if nums[left] == nums[right]:
                left += 1
                continue
            if nums[mid] <= nums[right]:
                if nums[mid] < target and nums[right] >= target:
                    left = mid+1
                else:
                    right = mid-1
            else:
                if nums[mid] > target and nums[left] <= target:
                    right = mid-1
                else:
                    left = mid+1

        if nums[left] == target: return True
        else: return False